File: C_code/VAD.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <sndfile.hh>

// Function to compute the energy of an audio signal frame-by-frame
std::vector<double> computeFrameEnergy(const std::vector<double>& signal, int frameLength, int hopLength) {
    std::vector<double> energy;
    for(size_t i = 0; i < signal.size(); i += hopLength) {
        double sum = 0.0;
        for(size_t j = i; j < i + frameLength && j < signal.size(); ++j) {
            sum += signal[j] * signal[j];
        }
        energy.push_back(sum);
    }
    return energy;
}

// Function for a simple voice activity detection (VAD)
std::vector<bool> vad(const std::vector<double>& signal, int frameLength, int hopLength, double energyThreshold) {
    // Normalize signal
    double maxVal = *std::max_element(signal.begin(), signal.end());
    std::vector<double> normalizedSignal(signal.size());
    std::transform(signal.begin(), signal.end(), normalizedSignal.begin(),
                   [&maxVal](double val) { return val / maxVal; });

    // Compute the energy of audio frames
    std::vector<double> energy = computeFrameEnergy(normalizedSignal, frameLength, hopLength);

    // Normalize energy
    double maxEnergy = *std::max_element(energy.begin(), energy.end());
    for (double & e : energy) {
        e /= maxEnergy;
    }

    // Detect voice activity
    std::vector<bool> vadResult(energy.size());
    std::transform(energy.begin(), energy.end(), vadResult.begin(),
                   [energyThreshold](double e) { return e > energyThreshold; });

    return vadResult;
}

// Function to isolate speech
std::vector<double> isolateSpeech(const std::vector<double>& signal, int frameLength, int hopLength, double energyThreshold) {
    std::vector<bool> vadResult = vad(signal, frameLength, hopLength, energyThreshold);
    std::vector<double> speechSignal(signal.size(), 0.0);

    for(size_t i = 0; i < vadResult.size(); i++) {
        if(vadResult[i]) {
            size_t start = i * hopLength;
            size_t end = std::min(start + frameLength, signal.size());
            std::copy(signal.begin() + start, signal.begin() + end, speechSignal.begin() + start);
        }
    }

    return speechSignal;
}

int main() {
    // Open the WAV file
    const char* filePath = "test.wav";
    SF_INFO sfInfo;
    SNDFILE* inFile = sf_open(filePath, SFM_READ, &sfInfo);

    if (inFile == nullptr) {
        std::cout << "Cannot open input file!\n";
        return 1;
    }

    // Read samples
    std::vector<double> signal(sfInfo.frames);
    sf_read_double(inFile, &signal[0], sfInfo.frames);
    sf_close(inFile);

    // Isolate speech from the signal
    std::vector<double> isolatedSpeech = isolateSpeech(signal, 1024, 512, 0.01);

    // Save the isolated speech to a new WAV file
    const char* outputFilePath = "isolated_speech_output.wav";
    SF_INFO sfOutInfo = sfInfo; // Copy input SF_INFO for output file
    SNDFILE* outFile = sf_open(outputFilePath, SFM_WRITE, &sfOutInfo);

    if (outFile == nullptr) {
        std::cout << "Cannot open output file!\n";
        return 1;
    }

    sf_write_double(outFile, &isolatedSpeech[0], sfOutInfo.frames);
    sf_close(outFile);

    std::cout << "Isolated speech saved to " << outputFilePath << std::endl;
    return 0;
}

File: FastICA/vad.py
import librosa
import numpy as np
import soundfile as sf

def compute_frame_energy(signal, frame_length, hop_length):
    # Compute the energy of an audio signal frame-by-frame
    energy = np.array([
        np.sum(np.abs(signal[i:i+frame_length])**2)
        for i in range(0, len(signal), hop_length)
    ])
    return energy

def vad(signal, sr, frame_length=1024, hop_length=512, energy_threshold=0.01):
    # Normalize signal
    signal = signal / np.max(np.abs(signal))

    # Compute the energy of audio frames
    energy = compute_frame_energy(signal, frame_length, hop_length)

    # Normalize energy
    energy = energy / np.max(np.abs(energy))

    # Detect voice activity
    vad_result = energy > energy_threshold

    return vad_result

def isolate_speech(signal, sr, frame_length=1024, hop_length=512, energy_threshold=0.01):
    # Perform VAD
    vad_result = vad(signal, sr, frame_length, hop_length, energy_threshold)

    # Initialize an empty signal for speech
    speech_signal = np.zeros_like(signal)

    # Iterate over frames and copy only speech frames to speech_signal
    for i, is_speech in enumerate(vad_result):
        start = i * hop_length
        end = start + frame_length
        if is_speech:
            speech_signal[start:end] = signal[start:end]

    return speech_signal

# Load an audio file
file_path = 'test.wav'
signal, sr = librosa.load(file_path, sr=None)

# Isolate speech from the signal
isolated_speech = isolate_speech(signal, sr)

# Save the modified signal to a new WAV file
output_file_path = 'isolated_speech_output.wav'  # Define your output file name
sf.write(output_file_path, isolated_speech, sr)

print(f"Isolated speech saved to {output_file_path}")
